File: archives/C#/0138. Copy List with Random Pointer.cs
/*
// Definition for a Node.
public class Node {
    public int val;
    public Node next;
    public Node random;

    public Node(){}
    public Node(int _val,Node _next,Node _random) {
        val = _val;
        next = _next;
        random = _random;
}
*/
public class Solution {
    public Node CopyRandomList(Node head) {
        if(head==null){
            return head;
        }
        Node node=head;
        while(node!=null){
            Node tmp=node.next;
            Node newnode=new Node(node.val,null,null);
            newnode.next=tmp;
            node.next=newnode;
            node=node.next.next;
        }
        
        Node pNode=head;
        while(pNode!=null){
            if(pNode.random!=null){
                pNode.next.random=pNode.random.next;
            }
            pNode=pNode.next.next;
        }
        
        Node oldNode=head;
        Node newNode=head.next;
        Node pold=oldNode;
        Node pnew=newNode;
        
        while(pnew.next!=null){
            pold.next=pnew.next;
            pold=pold.next;
            pnew.next=pold.next;
            pnew=pnew.next;
        }
        pnew.next=null;
        pold.next=null;
        return newNode;
        
    }
}

File: archives/20190519python/0138. Copy List with Random Pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, next, random):
        self.val = val
        self.next = next
        self.random = random
"""
class Solution:
    def copyRandomList(self, head: 'Node') -> 'Node':
        if not head:
            return None
        
        p=head
        while p:
            node=Node(p.val,None,None)
            node.next=p.next
            p.next=node
            p=p.next.next
            
        p=head
        while p:
            if p.random:
                p.next.random=p.random.next
            p=p.next.next
            
        oldhead=head
        newhead=head.next
        p_old=oldhead
        p_new=newhead
        while p_new.next:
            p_old.next=p_old.next.next
            p_old=p_old.next
            p_new.next=p_new.next.next
            p_new=p_new.next
        p_old.next=None
        return newhead
